### PR TITLE
Add prefix to declarative config resources ID

### PR DESCRIPTION
**Pull request by ivan-degtiarenko**
## Description

Based on this discussion: https://github.com/stackrox/stackrox/pull/6172#discussion_r1206101388
TLDR integration health of integration need to have different ID from declarative config health - that's why the prefix is added.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

TODO(replace-me)
Use this space to explain how you tested your PR, or, if you didn't test it, why
you did not do so. Valid reasons include, for example, "CI is sufficient",
"No testable changes". Feel free to attach JSON snippets, curl commands,
screenshots.

In addition to reviewing your code, reviewers **must** also review your testing
instructions and make sure they are sufficient.

### CODE REVIEW

**Comment by ivan-degtiarenko** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6172** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6172" target="_blank">Graphite</a> 
    * **PR #6184** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6184" target="_blank">Graphite</a> 
      * **PR #6198** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6198" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/6198?utm_source=stack-comment).

**Review by janisz — CHANGES_REQUESTED** [review]
Why do we need this prefix?

File: central/declarativeconfig/utils/integration_health.go [hunk 1]
```diff
@@ -12,13 +12,16 @@ import (
 	"github.com/stackrox/rox/pkg/utils"
 )
 
+// DeclarativeConfigHealthIDPrefix is prefix of all integration health's IDs of declarative configurations.
+const DeclarativeConfigHealthIDPrefix = "declarativeconfig."
```
Comment (janisz) [inline central/declarativeconfig/utils/integration_health.go:16]: what's the current format of ID? I think we discussed some time ago to switch to UUIDs everywhere and use this type on database. With string prefix it will not be possible? Why not add a field that will explicitly indicate origin?

**Review by janisz — COMMENTED** [review]
Could you add a test that will check the case Daniel pointed out?

File: central/declarativeconfig/manager_impl.go [hunk 2]
```diff
@@ -91,7 +91,7 @@ func New(reconciliationTickerDuration, watchIntervalDuration time.Duration, upda
 	writeDeclarativeRoleCtx = sac.WithGlobalAccessScopeChecker(writeDeclarativeRoleCtx,
 		sac.AllowFixedScopes(
 			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS, storage.Access_READ_WRITE_ACCESS),
-			sac.ResourceScopeKeys(resources.Access)))
+			sac.ResourceScopeKeys(resources.Access, resources.Integration)))
```
Comment (dhaus67) [inline central/declarativeconfig/manager_impl.go:94]: wait why do we need integration now here? something that was missing beforehand?

File: central/declarativeconfig/types/extractors.go [hunk 3]
```diff
@@ -15,12 +15,12 @@ type IDExtractor func(m proto.Message) string
 // NameExtractor extracts the name from proto messages.
 type NameExtractor func(m proto.Message) string
 
-// UniversalIDExtractor provides a way to extract the ID from proto messages.
+// UniversalIDExtractor provides a way to extractIDFromProtoMessage the ID from proto messages.
```
Comment (dhaus67) [inline central/declarativeconfig/types/extractors.go:18]: That's probably not wanted.

File: central/declarativeconfig/manager_impl.go [hunk 4]
```diff
@@ -375,11 +375,14 @@ func (m *managerImpl) removeStaleHealthStatuses(idsToSkip []string) error {
 		return errors.Wrap(err, "retrieving integration health statuses for declarative config")
 	}
 
-	idsToSkipSet := set.NewFrozenStringSet(idsToSkip...)
+	healthIDsToSkip := set.NewStringSet()
+	for _, id := range idsToSkip {
+		healthIDsToSkip.Add(declarativeConfigUtils.DeclarativeConfigHealthIDPrefix + id)
```
Comment (dhaus67) [inline central/declarativeconfig/manager_impl.go:380]: Since we already have the helper defined within the tests (`addHealthIDPrefix`), can we add a helper within the `declarativeConfigUtils` package that has the following signature:
```go
func IDForIntegrationHealth(id string) string
```

You can either skip changing all the IDs in the existing string set and within the `Contains` call `IDForIntegrationHealth` or remove the current prefix adding.

Also, we can then make `DeclarativeConfigHealthIDPrefix` unexposed, since there's no good reason to have it exposed anymore.

File: central/declarativeconfig/utils/integration_health.go (hunk 1 above)
Comment (dhaus67) [reply central/declarativeconfig/utils/integration_health.go:16]: Currently it's not constrained into any type of format: https://github.com/stackrox/stackrox/blob/master/proto/storage/integration_health.proto#L25

I do remember this as well. I think the issue here is that currently the ID of the proto will be taken as PK. In case of notifiers, we will potentially override the notifier (with the introduction of the new declarative config type, probably the ID shouldn't be the PK but rather the constraint of ID + integration health type). This would probably be the best solution, this way it would also be possible in the future to switch to `type: uuid` in the proto without huge changes required within the migration.

I'm not sure how easy it currently is to create constraints on the store, I had to do one for groups and was required to go through some hoops: #5099 (and the follow-up PR there). Not sure some of the previous issues were resolved or not.

I don't think it's an option to add the origin to the integration health proto itself, the ID issue would still exist then. So, either we take the prefix, knowing that in the future once the switch to `type:uuid` may come, we have to face the problem again and most likely go for the unique constraint on ID + integration health type, or we do this from the start.

Seeing the options, I'd prefer to go with the unique constraint since this one will be future proof and we won't have to do the workaround we currently do within the manager's implementation (i.e. prefixing the UUID and handling the edge cases there). I think we also have enough time to get this in 4.1. If we won't make it, then the prefix approach should be straightforward.

WDYT @ivan-degtiarenko ?

**Comment by ivan-degtiarenko** [comment]
Closing as we decided to go separate datastore route: https://github.com/stackrox/stackrox/pull/6317
